src/browser/serial.js

```diff
@@ -42,28 +44,59 @@ function SerialAdapter(element, bus)
     };
     this.init();
 
-
     this.show_char = function(chr)
     {
-        if(chr === "\x08")
-        {
-            this.text = this.text.slice(0, -1);
-            this.update();
-        }
-        else if(chr === "\r")
+        if(this.control_mode === false)
         {
-            // do nothing
+            if(chr === "\x08")
+            {
+                this.text = this.text.slice(0, -1);
+                this.update();
+            }
+            else if(chr === "\x1b") {
+               this.control_mode = true;
+               this.control_buffer = "";
+            }
+            else if(chr === "\r")
+            {
+                // do nothing
+            }
+            else
+            {
+               this.text += chr;
+
+               if(chr === "\n")
+               {
+                  this.text_new_line = true;
+               }
+
+               this.update();
+            }
         }
         else
         {
-            this.text += chr;
-
-            if(chr === "\n")
+            this.control_buffer += chr;
+            var cmds = {
+                bksp: /^\[[0-2]?J$/,
+                clr: /^\[[0-9;]{0,5}H\x1b\[[0-2]?J$/,
+                invalid: /[hl=>0-2M-OmrA-Efg3-8KnRc8qy]/
```

Comment:
This is really hard to understand, can you find another way to find the end of the control buffer? If not, I can give this a shot.

Maybe a single regular expression could be used to match any sequence as described [here](https://en.wikipedia.org/wiki/ANSI_escape_code#CSI_sequences):

> The ESC [ is followed by any number (including none) of "parameter bytes" in the range 0x30–0x3F (ASCII 0–9:;<=>?), then by any number of "intermediate bytes" in the range 0x20–0x2F (ASCII space and !"#$%&'()*+,-./), then finally by a single "final byte" in the range 0x40–0x7E (ASCII @A–Z[\]^_`a–z{|}~).

Reply:
We could go off of that list. I created that for compatibility with VT-100 controls as well. If need be I will fix it.

Reply:
I might just wait for this PR to go through and write a new version of this file with screen memory so I can implement more commands